src/vs/workbench/contrib/remote/browser/remoteStartEntry.ts

```diff
@@ -75,63 +91,116 @@ export class RemoteStartEntry extends Disposable implements IWorkbenchContributi
 			for (const ext of result) {
-				if (ext.identifier.id === this.currentRemoteExtensionId) {
-					this.registerRemoteCommand(ext.local?.manifest?.contributes, this.currentRemoteExtensionId);
+				const index = this.remoteExtensionMetadata.findIndex(value => value.id === ext.identifier.id && !value.installed);
+				if (index > -1) {
+					const command = await this.getRemoteCommand(ext.identifier.id);
+					if (command) {
+						this.remoteExtensionMetadata[index].remoteCommand = command;
+						this.remoteExtensionMetadata[index].installed = true;
+					}
 				}
 			}
 		}));
 
-		this._register(this.extensionManagementService.onDidUninstallExtension(result => {
-			if (this.remoteCommands.has(result.identifier.id)) {
-				this.remoteCommands.delete(result.identifier.id);
+		this._register(this.extensionManagementService.onDidUninstallExtension(async (result) => {
+			const index = this.remoteExtensionMetadata.findIndex(value => value.id === result.identifier.id && value.installed);
+			if (index > -1) {
+				this.remoteExtensionMetadata[index].installed = false;
+				if (this.remoteExtensionMetadata[index].dependenciesStr === undefined) {
+					const dependenciesStr = await this.getDependenciesStr(this.remoteExtensionMetadata[index].id);
+					this.remoteExtensionMetadata[index].dependenciesStr = dependenciesStr;
+				}
 			}
 		}));
 	}
 
-	private registerRemoteCommand(extensionContributions: IExtensionContributions | undefined, remoteId: string) {
-		if (extensionContributions?.menus) {
-			for (const context in extensionContributions.menus) {
-				if (context === STATUSBAR_REMOTEINDICATOR_CONTRIBUTION) {
-					const command = extensionContributions.menus[context][0];
-					this.remoteCommands.set(remoteId, command.command);
+	private async _init(): Promise<void> {
+
+		if (this._isInitialized) {
+			return;
+		}
+
+		for (let i = 0; i < this.remoteExtensionMetadata.length; i++) {
+			const installed = this.extensionService.extensions.some((e) => e.id?.toLowerCase() === this.remoteExtensionMetadata[i].id);
+			if (installed) {
+				const command = await this.getRemoteCommand(this.remoteExtensionMetadata[i].id);
+				if (command) {
+					this.remoteExtensionMetadata[i].remoteCommand = command;
+					this.remoteExtensionMetadata[i].installed = true;
+				}
+			}
+			else {
+				const dependenciesStr = await this.getDependenciesStr(this.remoteExtensionMetadata[i].id);
+				this.remoteExtensionMetadata[i].dependenciesStr = dependenciesStr;
+			}
+		}
+
+		this._isInitialized = true;
+	}
+
+	private async getRemoteCommand(remoteExtensionId: string): Promise<string | undefined> {
+
+		// Remote extension is not registered immediately on install.
+		// Wait for it to appear before returning.
+		let extension: Readonly<IRelaxedExtensionDescription> | undefined;
+		while (!extension) {
+			extension = await this.extensionService.getExtension(remoteExtensionId);
+			await timeout(300);
```

Comment:
Would something like this be better? https://github.com/microsoft/vscode/blob/7dfc1938832b0a91aec38b079a0f3cf3e404f980/src/vs/workbench/browser/layout.ts#L844

Reply:
Thanks. I used the getExtension since I'm using that object down below anyway. 
I changed up the logic to use retry to do this instead. 

https://github.com/microsoft/vscode/blob/f708bbe9f2f07fca45533e1c5a7afc23a48746aa/src/vs/base/common/async.ts#L1234